File: frontend/mobile_web_flutter/lib/core/dashboard_service.dart
// lib/core/dashboard_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;

import 'api_base.dart';

class DetectionTimePoint {
  final DateTime date;
  final int count;

  DetectionTimePoint({required this.date, required this.count});

  factory DetectionTimePoint.fromJson(Map<String, dynamic> json) {
    return DetectionTimePoint(
      date: DateTime.parse(json['date'] as String),
      count: json['count'] as int? ?? 0,
    );
  }
}

class DiseaseStat {
  final String diseaseName;
  final int count;

  DiseaseStat({required this.diseaseName, required this.count});

  factory DiseaseStat.fromJson(Map<String, dynamic> json) {
    return DiseaseStat(
      diseaseName: json['disease_name'] as String? ?? 'Không rõ',
      count: json['count'] as int? ?? 0,
    );
  }
}

class TicketStatusStat {
  final String status;
  final int count;

  TicketStatusStat({required this.status, required this.count});

  factory TicketStatusStat.fromJson(Map<String, dynamic> json) {
    return TicketStatusStat(
      status: json['status'] as String? ?? 'unknown',
      count: json['count'] as int? ?? 0,
    );
  }
}

class RecentDetectionItem {
  final int detectionId;
  final int? userId;
  final String? username;
  final String? diseaseName;
  final double? confidence;
  final DateTime createdAt;

  RecentDetectionItem({
    required this.detectionId,
    this.userId,
    this.username,
    this.diseaseName,
    this.confidence,
    required this.createdAt,
  });

  factory RecentDetectionItem.fromJson(Map<String, dynamic> json) {
    return RecentDetectionItem(
      detectionId: json['detection_id'] as int,
      userId: json['user_id'] as int?,
      username: json['username'] as String?,
      diseaseName: json['disease_name'] as String?,
      confidence: json['confidence'] == null
          ? null
          : (json['confidence'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class RecentTicketItem {
  final int ticketId;
  final int? userId;
  final String? username;
  final String? status;
  final String? title;
  final DateTime createdAt;

  RecentTicketItem({
    required this.ticketId,
    this.userId,
    this.username,
    this.status,
    this.title,
    required this.createdAt,
  });

  factory RecentTicketItem.fromJson(Map<String, dynamic> json) {
    return RecentTicketItem(
      ticketId: json['ticket_id'] as int,
      userId: json['user_id'] as int?,
      username: json['username'] as String?,
      status: json['status'] as String?,
      title: json['title'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class DashboardSummary {
  final int totalDevices;
  final int activeDevices;
  final int inactiveDevices;
  final int totalUsers;
  final int newUsers;
  final int totalDetections;
  final List<DetectionTimePoint> detectionsOverTime;
  final List<DiseaseStat> topDiseases;
  final int totalTickets;
  final int openTickets;
  final List<TicketStatusStat> ticketsByStatus;
  final List<RecentDetectionItem> recentDetections;
  final List<RecentTicketItem> recentTickets;

  DashboardSummary({
    required this.totalDevices,
    required this.activeDevices,
    required this.inactiveDevices,
    required this.totalUsers,
    required this.newUsers,
    required this.totalDetections,
    required this.detectionsOverTime,
    required this.topDiseases,
    required this.totalTickets,
    required this.openTickets,
    required this.ticketsByStatus,
    required this.recentDetections,
    required this.recentTickets,
  });

  factory DashboardSummary.fromJson(Map<String, dynamic> json) {
    return DashboardSummary(
      totalDevices: json['total_devices'] as int? ?? 0,
      activeDevices: json['active_devices'] as int? ?? 0,
      inactiveDevices: json['inactive_devices'] as int? ?? 0,
      totalUsers: json['total_users'] as int? ?? 0,
      newUsers: json['new_users'] as int? ?? 0,
      totalDetections: json['total_detections'] as int? ?? 0,
      detectionsOverTime:
          (json['detections_over_time'] as List<dynamic>? ?? [])
              .map((e) => DetectionTimePoint.fromJson(
                  Map<String, dynamic>.from(e as Map)))
              .toList(),
      topDiseases: (json['top_diseases'] as List<dynamic>? ?? [])
          .map(
              (e) => DiseaseStat.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      totalTickets: json['total_tickets'] as int? ?? 0,
      openTickets: json['open_tickets'] as int? ?? 0,
      ticketsByStatus:
          (json['tickets_by_status'] as List<dynamic>? ?? [])
              .map((e) => TicketStatusStat.fromJson(
                  Map<String, dynamic>.from(e as Map)))
              .toList(),
      recentDetections:
          (json['recent_detections'] as List<dynamic>? ?? [])
              .map((e) => RecentDetectionItem.fromJson(
                  Map<String, dynamic>.from(e as Map)))
              .toList(),
      recentTickets: (json['recent_tickets'] as List<dynamic>? ?? [])
          .map((e) => RecentTicketItem.fromJson(
              Map<String, dynamic>.from(e as Map)))
          .toList(),
    );
  }
}

class DashboardService {
  /// range: '7d' | '30d' | '90d'
  static Future<DashboardSummary> fetchSummary({String range = '7d'}) async {
    final params = <String, String>{
      'range': range,
    };

    final query = Uri(queryParameters: params).query;

    final res = await ApiBase.getJson(
      ApiBase.api('/admin/dashboard?$query'),
    );

    final map = Map<String, dynamic>.from(res as Map);
    return DashboardSummary.fromJson(map);
  }

  /// ADMIN: GET /api/v1/admin/dashboard/export?range=
  /// Xuất báo cáo dashboard dạng CSV (nếu bạn vẫn muốn giữ)
  static Future<String> exportDashboardCsv({String range = '7d'}) async {
    final params = <String, String>{
      'range': range,
    };

    final query = Uri(queryParameters: params).query;

    final uri = Uri.parse(
      '${ApiBase.baseURL}${ApiBase.api('/admin/dashboard/export?$query')}',
    );

    final token = ApiBase.bearer;
    final headers = <String, String>{
      'Accept': 'text/csv',
      if (token != null && token.isNotEmpty) 'Authorization': 'Bearer $token',
    };

    final resp = await http.get(uri, headers: headers);

    if ((resp.statusCode ~/ 100) != 2) {
      throw Exception(
        'Xuất báo cáo dashboard (CSV) thất bại (${resp.statusCode}): ${resp.body}',
      );
    }

    return utf8.decode(resp.bodyBytes);
  }

  /// ADMIN: GET /api/v1/admin/reports/summary?range=
  /// Xuất báo cáo tổng quan dạng PDF
  static Future<Uint8List> exportSummaryPdf({String range = '7d'}) async {
    final params = <String, String>{
      'range': range,
    };

    final query = Uri(queryParameters: params).query;

    final uri = Uri.parse(
      '${ApiBase.baseURL}${ApiBase.api('/admin/reports/summary?$query')}',
    );

    final token = ApiBase.bearer;
    final headers = <String, String>{
      'Accept': 'application/pdf',
      if (token != null && token.isNotEmpty) 'Authorization': 'Bearer $token',
    };

    final resp = await http.get(uri, headers: headers);

    if ((resp.statusCode ~/ 100) != 2) {
      throw Exception(
        'Xuất báo cáo PDF thất bại (${resp.statusCode}): ${resp.body}',
      );
    }

    return resp.bodyBytes;
  }
}

File: backend/app/services/dashboard_service.py
from datetime import datetime, timedelta
from typing import List

from sqlalchemy.orm import Session
from sqlalchemy import func, desc

from app.models.devices import Device
from app.models.user import Users
from app.models.image_detection import Img, Detection, Disease
from app.models.support import SupportTicket  # model mapping bảng support_tickets

from app.schemas.dashboard import (
    DashboardSummary,
    DetectionTimePoint,
    DiseaseStat,
    TicketStatusStat,
    RecentDetectionItem,
    RecentTicketItem,
)


def _get_range_dates(range_str: str) -> tuple[datetime, datetime]:
    """
    range_str: '7d' | '30d' | '90d'
    """
    now = datetime.utcnow()
    if range_str == "30d":
        start = now - timedelta(days=30)
    elif range_str == "90d":
        start = now - timedelta(days=90)
    else:  # default 7d
        start = now - timedelta(days=7)
    return start, now


def build_dashboard_summary(db: Session, range_str: str = "7d") -> DashboardSummary:
    start, end = _get_range_dates(range_str)

    # ====================== DEVICES ======================
    # devices: device_id, status ENUM('active','maintain','inactive')
    total_devices = db.query(func.count(Device.device_id)).scalar() or 0

    active_devices = (
        db.query(func.count(Device.device_id))
        .filter(Device.status == "active")
        .scalar()
        or 0
    )
    inactive_devices = (
        db.query(func.count(Device.device_id))
        .filter(Device.status == "inactive")
        .scalar()
        or 0
    )
    # (tùy bạn có dùng hay không, maintain nằm ngoài 2 nhóm trên)

    # ======================== USERS ======================
    total_users = db.query(func.count(Users.user_id)).scalar() or 0
    # users.created_at: TIMESTAMP DEFAULT CURRENT_TIMESTAMP
    new_users = (
        db.query(func.count(Users.user_id))
        .filter(Users.created_at >= start)
        .scalar()
        or 0
    )

    # ===================== DETECTIONS ====================
    # detections: detection_id, img_id, disease_id, confidence, created_at
    detections_range_q = (
        db.query(Detection)
        .filter(Detection.created_at >= start, Detection.created_at <= end)
    )

    total_detections = detections_range_q.count()

    # detections_over_time: group by DATE(created_at)
    detections_over_time_rows = (
        db.query(
            func.date(Detection.created_at).label("d"),
            func.count(Detection.detection_id),
        )
        .filter(Detection.created_at >= start, Detection.created_at <= end)
        .group_by(func.date(Detection.created_at))
        .order_by(func.date(Detection.created_at))
        .all()
    )
    detections_over_time: List[DetectionTimePoint] = [
        DetectionTimePoint(date=row[0], count=row[1])
        for row in detections_over_time_rows
    ]

    # top_diseases: group by disease_id (join bảng diseases)
    top_diseases_rows = (
        db.query(
            Disease.name,
            func.count(Detection.detection_id).label("cnt"),
        )
        .join(Detection, Detection.disease_id == Disease.disease_id)
        .filter(Detection.created_at >= start, Detection.created_at <= end)
        .group_by(Disease.name)
        .order_by(desc("cnt"))
        .limit(5)
        .all()
    )
    top_diseases: List[DiseaseStat] = [
        DiseaseStat(disease_name=name or "Không rõ", count=cnt)
        for name, cnt in top_diseases_rows
    ]

    # ====================== TICKETS ======================
    # support_tickets: ticket_id, user_id, title, description,
    # status ENUM('processing','processed'), created_at
    tickets_range_q = (
        db.query(SupportTicket)
        .filter(SupportTicket.created_at >= start,
                SupportTicket.created_at <= end)
    )

    total_tickets = tickets_range_q.count()

    # "open_tickets" = ticket đang xử lý (processing)
    open_tickets = (
        tickets_range_q.filter(SupportTicket.status == "processing").count()
    )

    tickets_by_status_rows = (
        db.query(
            SupportTicket.status,
            func.count(SupportTicket.ticket_id),
        )
        .filter(SupportTicket.created_at >= start,
                SupportTicket.created_at <= end)
        .group_by(SupportTicket.status)
        .all()
    )
    tickets_by_status: List[TicketStatusStat] = [
        TicketStatusStat(status=status or "unknown", count=count)
        for status, count in tickets_by_status_rows
    ]

    # ============= RECENT DETECTIONS (10 GẦN NHẤT) =============
    recent_detection_rows = (
        db.query(Detection, Img, Users, Disease)
        .join(Img, Detection.img_id == Img.img_id)
        .outerjoin(Users, Img.user_id == Users.user_id)
        .outerjoin(Disease, Detection.disease_id == Disease.disease_id)
        .order_by(desc(Detection.created_at))
        .limit(10)
        .all()
    )

    recent_detections: List[RecentDetectionItem] = []
    for det, img, user, disease in recent_detection_rows:
        recent_detections.append(
            RecentDetectionItem(
                detection_id=det.detection_id,
                user_id=user.user_id if user else None,
                username=user.username if user else None,
                disease_name=disease.name if disease else None,
                confidence=float(det.confidence)
                if det.confidence is not None
                else None,
                created_at=det.created_at,
            )
        )

    # ============= RECENT TICKETS (10 GẦN NHẤT) =============
    recent_tickets_rows = (
        db.query(SupportTicket, Users)
        .outerjoin(Users, SupportTicket.user_id == Users.user_id)
        .order_by(desc(SupportTicket.created_at))
        .limit(10)
        .all()
    )

    recent_tickets: List[RecentTicketItem] = []
    for ticket, user in recent_tickets_rows:
        recent_tickets.append(
            RecentTicketItem(
                ticket_id=ticket.ticket_id,
                user_id=user.user_id if user else None,
                username=user.username if user else None,
                status=ticket.status,
                title=ticket.title,
                created_at=ticket.created_at,
            )
        )

    # ====================== KẾT QUẢ ======================
    return DashboardSummary(
        total_devices=total_devices,
        active_devices=active_devices,
        inactive_devices=inactive_devices,
        total_users=total_users,
        new_users=new_users,
        total_detections=total_detections,
        detections_over_time=detections_over_time,
        top_diseases=top_diseases,
        total_tickets=total_tickets,
        open_tickets=open_tickets,
        tickets_by_status=tickets_by_status,
        recent_detections=recent_detections,
        recent_tickets=recent_tickets,
    )
